mapping_parenting_tech/analysis/prototyping/play_store/get_play_store_apps.py

```diff
@@ -7,141 +7,413 @@
-#       jupytext_version: 1.13.3
+#       jupytext_version: 1.13.4
 #   kernelspec:
 #     display_name: 'Python 3.8.12 64-bit (''mapping_parenting_tech'': conda)'
 #     language: python
 #     name: python3
 # ---
 
 # %% [markdown]
-# # Code to extract app ids from [list of] web pages saved from Google Play store
+# # Extract app ids from web pages saved from Google Play store
+
+# %% [markdown]
+# Code below carries out the following set of functions:
+# 1. retrieves the ids of apps from a Play Store web page that has been saved locally
+# 2. saves and loads app ids to/from a pickle file so they can be retrieved and used later
+# 3. downloads details for an app/s using its/their app id/s
+# 4. saves and loads app details so they can be retrieved and used later
+# 5.
 
 # %% [markdown]
 # Do imports and set file locations
 
 # %%
-from google_play_scraper import Sort, app, reviews_all
-from tqdm import tqdm
 import re
 import json
+import pickle
+import csv
+from google_play_scraper import Sort, app, reviews
+from tqdm import tqdm
+from pathlib import Path
 from mapping_parenting_tech import PROJECT_DIR, logging
 
-INPUT_PATH = PROJECT_DIR / "inputs/data"
-OUTPUT_PATH = PROJECT_DIR / "outputs/data"
-STORE_PATH = "play_store/"
-PLAY_DATA_FILE = "playstore_innerHTML.html"
-OUTPUT_PLAY_IDS_FILE = "play_store_ids.json"
-OUTPUT_PLAY_DETAILS_FILE = "play_store_details.json"
-OUTPUT_PLAY_REVIEWS_FILE = "play_store_reviews.json"
 
-input_data = INPUT_PATH / STORE_PATH / PLAY_DATA_FILE
+# %% [markdown]
+# ## Read HTML and extract list of app ids
+
+# %%
+def get_playstore_app_ids(filename: str) -> list:
+    """
+    Looks inside 'inputs/data' int the project directory for a given HTML file (filename) and scans that file
+    for links containing ids to Play Store apps.
+
+    Argument:
+        filename: str - the file name of the HTML file you wish to get app ids from. Must be in <[project directory]/inputs/data/>
+
+    Returns:
+        List of app ids, each as a str
+    """
+
+    # open the input file
+    with open(PROJECT_DIR / "inputs/data" / filename, "rt") as infile:
+        html = infile.read()
 
-if input_data.exists():
-    logging.info("Ok to proceed")
-else:
-    logging.warning("Didn't find file")
+    # setup the regular expression that looks for app ids in the links to apps on a category page
+    # (e.g., https://play.google.com/store/apps/collection/cluster?clp=0g4hCh8KGXRvcHNlbGxpbmdfZnJlZV9QQVJFTlRJTkcQBxgD:S:ANO1ljI8w1M&gsr=CiTSDiEKHwoZdG9wc2VsbGluZ19mcmVlX1BBUkVOVElORxAHGAM%3D:S:ANO1ljK7gT4)
+    # this is looking for a link ('href') to '/store/apps/details?id=' and it grabs the text after 'id'
+    re_pattern = r"(?<=href=\"\/store\/apps\/details\?id=)(.*?)(?=\")"
+
+    # retrieve all RE matches - this will return duplicates
+    link_targets = re.findall(re_pattern, html)
+
+    # convert list of links into a dict to remove duplicates, and back into a list
+    app_ids = list(dict.fromkeys(link_targets))
```

Comment:
Using `set` to deduplicate a list is a bit more efficient according to [this discussion](https://stackoverflow.com/questions/60062818/difference-between-removing-duplicates-from-a-list-using-dict-and-set)
```suggestion
    app_ids = list(set(link_targets))
```

Reply:
Complete